File: scripts/test-api.mjs
/**
 * test-api.mjs
 * Tests all four backend API routes: /health, /check-env, /start-agent, /hangup-agent
 * Run via: node scripts/test-api.mjs
 */

const BASE_URL = "http://localhost:3000";

function pass(label) {
  console.log(`[PASS] ${label}`);
}

function fail(label, reason) {
  console.log(`[FAIL] ${label} — ${reason}`);
}

function info(label, data) {
  console.log(`[INFO] ${label}:`, JSON.stringify(data, null, 2));
}

async function testHealth() {
  console.log("\n--- GET /api/health ---");
  const res = await fetch(`${BASE_URL}/api/health`);
  const body = await res.json();
  info("Response", body);
  if (res.status === 200 && body.status === "ok") {
    pass("Health endpoint returns 200 with { status: 'ok' }");
    return true;
  } else {
    fail("Health endpoint", `Expected 200 + status:ok, got ${res.status} / ${JSON.stringify(body)}`);
    return false;
  }
}

async function testCheckEnv() {
  console.log("\n--- GET /api/check-env ---");
  const res = await fetch(`${BASE_URL}/api/check-env`);
  const body = await res.json();
  info("Response", body);

  if (res.status !== 200) {
    fail("check-env", `Expected HTTP 200, got ${res.status}`);
    return { ready: false };
  }

  pass("check-env returns HTTP 200");

  if (body.ready) {
    pass("All required env vars are configured");
  } else {
    fail("check-env", `Missing required env vars: ${JSON.stringify(body.missing)}`);
  }

  return body;
}

async function testStartAgent(envReady) {
  console.log("\n--- POST /api/start-agent ---");

  if (!envReady) {
    console.log("[SKIP] Skipping start-agent — env vars not ready");
    return null;
  }

  const res = await fetch(`${BASE_URL}/api/start-agent`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      prompt: "You are a test assistant. Respond with 'Test successful' to any message.",
      greeting: "Hello, I am the test agent.",
    }),
  });

  const body = await res.json();
  info("Response", body);

  if (res.status === 200 && body.success) {
    pass(`start-agent created agent (id: ${body.agentId}, channel: ${body.channel})`);
    return body;
  } else if (res.status === 502) {
    fail("start-agent", `Agora API rejected the request: ${JSON.stringify(body.error)}`);
    return null;
  } else {
    fail("start-agent", `HTTP ${res.status} — ${JSON.stringify(body)}`);
    return null;
  }
}

async function testHangupAgent(agentData) {
  console.log("\n--- POST /api/hangup-agent ---");

  if (!agentData) {
    console.log("[SKIP] Skipping hangup-agent — no agent session to hang up");
    return;
  }

  const { agentId, appId } = agentData;

  if (!agentId) {
    console.log("[SKIP] Skipping hangup-agent — agentId not returned from start-agent");
    return;
  }

  const res = await fetch(`${BASE_URL}/api/hangup-agent`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ agentId, appId }),
  });

  const body = await res.json().catch(() => ({}));
  info("Response", body);

  if (res.status === 200) {
    pass(`hangup-agent successfully stopped agent ${agentId}`);
  } else {
    fail("hangup-agent", `HTTP ${res.status} — ${JSON.stringify(body)}`);
  }
}

async function run() {
  console.log(`\n========================================`);
  console.log(`  Voice AI Agent — Backend API Test`);
  console.log(`  Target: ${BASE_URL}`);
  console.log(`========================================`);

  // 1. Health check
  const healthOk = await testHealth().catch((e) => {
    fail("health (connection error)", e.message);
    return false;
  });

  if (!healthOk) {
    console.log("\n[ERROR] Could not reach the server. Is 'pnpm dev' running on port 3000?");
    process.exit(1);
  }

  // 2. Check env vars
  const envStatus = await testCheckEnv().catch((e) => {
    fail("check-env (connection error)", e.message);
    return { ready: false };
  });

  // 3. Start agent (only if env is ready)
  const agentData = await testStartAgent(envStatus.ready).catch((e) => {
    fail("start-agent (connection error)", e.message);
    return null;
  });

  // 4. Hang up agent (only if start succeeded)
  await testHangupAgent(agentData).catch((e) => {
    fail("hangup-agent (connection error)", e.message);
  });

  console.log("\n========================================");
  console.log("  Test run complete.");
  console.log("========================================\n");
}

run();

File: scripts/test-api.py
import urllib.request
import urllib.error
import json
import time

BASE = "http://localhost:3000"

def request(method, path, body=None):
    url = BASE + path
    data = json.dumps(body).encode() if body else None
    headers = {"Content-Type": "application/json"}
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=10) as res:
            raw = res.read().decode()
            try:
                return res.status, json.loads(raw)
            except Exception:
                return res.status, raw
    except urllib.error.HTTPError as e:
        raw = e.read().decode()
        try:
            return e.code, json.loads(raw)
        except Exception:
            return e.code, raw
    except Exception as ex:
        return None, str(ex)

def check(label, status, body, expect_status=200):
    ok = status == expect_status
    icon = "PASS" if ok else "FAIL"
    print(f"[{icon}] {label}")
    print(f"       Status : {status}")
    if isinstance(body, dict):
        print(f"       Body   : {json.dumps(body, indent=2)}")
    else:
        print(f"       Body   : {body}")
    print()
    return ok

print("=" * 52)
print("  Voice AI Agent — Backend API Test Suite")
print("=" * 52)
print()

# 1. Health check
status, body = request("GET", "/api/health")
check("GET /api/health", status, body)

# 2. Check env
status, body = request("GET", "/api/check-env")
check("GET /api/check-env", status, body)

# 3. Start agent (expects the Agora ConvoAI call to work if env vars set)
start_payload = {
    "channel": "test-channel-123",
    "uid": 12345
}
status, body = request("POST", "/api/start-agent", start_payload)
# 200 = success, 500 = missing env vars or Agora error — both are informative
agent_id = None
if isinstance(body, dict) and body.get("agent_id"):
    agent_id = body["agent_id"]
check("POST /api/start-agent", status, body, expect_status=status)

# 4. Hangup agent (only if we got an agent_id back)
if agent_id:
    time.sleep(1)
    hangup_payload = {"agent_id": agent_id, "channel": "test-channel-123"}
    status, body = request("POST", "/api/hangup-agent", hangup_payload)
    check("POST /api/hangup-agent", status, body, expect_status=status)
else:
    print("[SKIP] POST /api/hangup-agent — no agent_id returned from start-agent")
    print()

print("=" * 52)
print("  Test run complete")
print("=" * 52)
